pkg/controller/instance/controller_reconcile.go

```diff
@@ -123,193 +155,117 @@ func (igr *instanceGraphReconciler) reconcileInstance(ctx context.Context) error
+		}
+		clusterObj, err := aset.Add(ctx, applyable)
+		if err != nil {
+			return fmt.Errorf("failed to add resource to applyset: %w", err)
+		}
 
-	// Check if resource exists
-	observed, err := rc.Get(ctx, resource.GetName(), metav1.GetOptions{})
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			// For read-only resources, we don't create
-			if igr.runtime.ResourceDescriptor(resourceID).IsExternalRef() {
-				resourceState.State = "WAITING_FOR_EXTERNAL_RESOURCE"
-				resourceState.Err = fmt.Errorf("external resource not found: %w", err)
-				return igr.delayedRequeue(resourceState.Err)
+		if clusterObj != nil {
+			igr.runtime.SetResource(resourceID, clusterObj)
+			igr.updateResourceReadiness(resourceID)
+			// Synchronize runtime state after each resource
+			if _, err := igr.runtime.Synchronize(); err != nil {
+				return fmt.Errorf("failed to synchronize after apply/prune: %w", err)
 			}
-			return igr.handleResourceCreation(ctx, rc, resource, resourceID, resourceState)
 		}
-		resourceState.State = ResourceStateError
-		resourceState.Err = fmt.Errorf("failed to get resource: %w", err)
-		return resourceState.Err
 	}
 
-	// Update runtime with observed state
-	igr.runtime.SetResource(resourceID, observed)
-
-	// Check resource readiness
-	if ready, reason, err := igr.runtime.IsResourceReady(resourceID); err != nil || !ready {
-		log.V(1).Info("Resource not ready", "reason", reason, "error", err)
-		resourceState.State = ResourceStateWaitingForReadiness
-		resourceState.Err = fmt.Errorf("resource not ready: %s: %w", reason, err)
-		return igr.delayedRequeue(resourceState.Err)
+	result, err := aset.Apply(ctx, prune)
+	for _, applied := range result.AppliedObjects {
+		resourceState := igr.state.ResourceStates[applied.ID]
+		if applied.Error != nil {
+			resourceState.State = ResourceStateError
+			resourceState.Err = applied.Error
+		} else {
+			igr.updateResourceReadiness(applied.ID)
+		}
 	}
 
-	resourceState.State = ResourceStateSynced
-
-	// For read-only resources, don't perform updates
-	if igr.runtime.ResourceDescriptor(resourceID).IsExternalRef() {
-		return nil
+	if err != nil {
+		return igr.delayedRequeue(fmt.Errorf("failed to apply/prune resources: %w", err))
 	}
 
-	return igr.updateResource(ctx, rc, resource, observed, resourceID, resourceState)
-}
+	// Inspect resource states and return error if any resource is in error state
+	if err := igr.state.ResourceErrors(); err != nil {
+		return igr.delayedRequeue(err)
+	}
 
-// getResourceClient returns the appropriate dynamic client and namespace for a resource
-func (igr *instanceGraphReconciler) getResourceClient(resourceID string) dynamic.ResourceInterface {
-	descriptor := igr.runtime.ResourceDescriptor(resourceID)
-	gvr := descriptor.GetGroupVersionResource()
-	namespace := igr.getResourceNamespace(resourceID)
+	if _, err := igr.runtime.Synchronize(); err != nil {
+		return fmt.Errorf("failed to synchronize after apply/prune: %w", err)
+	}
```

Comment:
Do we need to synchronize here? synchronize is only useful before applying new things.